// @azure/core-http/es/src/credentials/accessTokenRefresher.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { __awaiter, __generator } from "tslib";
/**
 * Helps the core-http token authentication policies with requesting a new token if we're not currently waiting for a new token.
 */
var AccessTokenRefresher = /** @class */ (function () {
    function AccessTokenRefresher(credential, scopes, requiredMillisecondsBeforeNewRefresh) {
        if (requiredMillisecondsBeforeNewRefresh === void 0) { requiredMillisecondsBeforeNewRefresh = 30000; }
        this.credential = credential;
        this.scopes = scopes;
        this.requiredMillisecondsBeforeNewRefresh = requiredMillisecondsBeforeNewRefresh;
        this.lastCalled = 0;
    }
    /**
     * Returns true if the required milliseconds(defaulted to 30000) have been passed signifying
     * that we are ready for a new refresh.
     *
     * @returns {boolean}
     */
    AccessTokenRefresher.prototype.isReady = function () {
        // We're only ready for a new refresh if the required milliseconds have passed.
        return (!this.lastCalled || Date.now() - this.lastCalled > this.requiredMillisecondsBeforeNewRefresh);
    };
    /**
     * Stores the time in which it is called,
     * then requests a new token,
     * then sets this.promise to undefined,
     * then returns the token.
     * @param options getToken options
     */
    AccessTokenRefresher.prototype.getToken = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var token;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.lastCalled = Date.now();
                        return [4 /*yield*/, this.credential.getToken(this.scopes, options)];
                    case 1:
                        token = _a.sent();
                        this.promise = undefined;
                        return [2 /*return*/, token || undefined];
                }
            });
        });
    };
    /**
     * Requests a new token if we're not currently waiting for a new token.
     * Returns null if the required time between each call hasn't been reached.
     * @param options getToken options
     */
    AccessTokenRefresher.prototype.refresh = function (options) {
        if (!this.promise) {
            this.promise = this.getToken(options);
        }
        return this.promise;
    };
    return AccessTokenRefresher;
}());
export { AccessTokenRefresher };
