// @azure/storage-blob/dist-esm/storage-internal-avro/src/AvroReader.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { __asyncGenerator, __await, __awaiter, __generator } from "tslib";
import { AVRO_SYNC_MARKER_SIZE, AVRO_INIT_BYTES, AVRO_CODEC_KEY, AVRO_SCHEMA_KEY } from "./AvroConstants";
import { arraysEqual } from "./utils/utils.common";
import { AvroType, AvroParser } from "./AvroParser";
import "@azure/core-paging";
var AvroReader = /** @class */ (function () {
    function AvroReader(dataStream, headerStream, currentBlockOffset, indexWithinCurrentBlock) {
        this._dataStream = dataStream;
        this._headerStream = headerStream || dataStream;
        this._initialized = false;
        this._blockOffset = currentBlockOffset || 0;
        this._objectIndex = indexWithinCurrentBlock || 0;
        this._initialBlockOffset = currentBlockOffset || 0;
    }
    Object.defineProperty(AvroReader.prototype, "blockOffset", {
        get: function () {
            return this._blockOffset;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AvroReader.prototype, "objectIndex", {
        get: function () {
            return this._objectIndex;
        },
        enumerable: false,
        configurable: true
    });
    AvroReader.prototype.initialize = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var header, _a, codec, _b, schema, _c, i;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0: return [4 /*yield*/, AvroParser.readFixedBytes(this._headerStream, AVRO_INIT_BYTES.length, {
                            abortSignal: options.abortSignal
                        })];
                    case 1:
                        header = _d.sent();
                        if (!arraysEqual(header, AVRO_INIT_BYTES)) {
                            throw new Error("Stream is not an Avro file.");
                        }
                        // File metadata is written as if defined by the following map schema:
                        // { "type": "map", "values": "bytes"}
                        _a = this;
                        return [4 /*yield*/, AvroParser.readMap(this._headerStream, AvroParser.readString, {
                                abortSignal: options.abortSignal
                            })];
                    case 2:
                        // File metadata is written as if defined by the following map schema:
                        // { "type": "map", "values": "bytes"}
                        _a._metadata = _d.sent();
                        codec = this._metadata[AVRO_CODEC_KEY];
                        if (!(codec == undefined || codec == "null")) {
                            throw new Error("Codecs are not supported");
                        }
                        // The 16-byte, randomly-generated sync marker for this file.
                        _b = this;
                        return [4 /*yield*/, AvroParser.readFixedBytes(this._headerStream, AVRO_SYNC_MARKER_SIZE, {
                                abortSignal: options.abortSignal
                            })];
                    case 3:
                        // The 16-byte, randomly-generated sync marker for this file.
                        _b._syncMarker = _d.sent();
                        schema = JSON.parse(this._metadata[AVRO_SCHEMA_KEY]);
                        this._itemType = AvroType.fromSchema(schema);
                        if (this._blockOffset == 0) {
                            this._blockOffset = this._initialBlockOffset + this._dataStream.position;
                        }
                        _c = this;
                        return [4 /*yield*/, AvroParser.readLong(this._dataStream, {
                                abortSignal: options.abortSignal
                            })];
                    case 4:
                        _c._itemsRemainingInBlock = _d.sent();
                        // skip block length
                        return [4 /*yield*/, AvroParser.readLong(this._dataStream, { abortSignal: options.abortSignal })];
                    case 5:
                        // skip block length
                        _d.sent();
                        this._initialized = true;
                        if (!(this._objectIndex && this._objectIndex > 0)) return [3 /*break*/, 9];
                        i = 0;
                        _d.label = 6;
                    case 6:
                        if (!(i < this._objectIndex)) return [3 /*break*/, 9];
                        return [4 /*yield*/, this._itemType.read(this._dataStream, { abortSignal: options.abortSignal })];
                    case 7:
                        _d.sent();
                        this._itemsRemainingInBlock--;
                        _d.label = 8;
                    case 8:
                        i++;
                        return [3 /*break*/, 6];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    AvroReader.prototype.hasNext = function () {
        return !this._initialized || this._itemsRemainingInBlock > 0;
    };
    AvroReader.prototype.parseObjects = function (options) {
        if (options === void 0) { options = {}; }
        return __asyncGenerator(this, arguments, function parseObjects_1() {
            var result, marker, _a, err_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!!this._initialized) return [3 /*break*/, 2];
                        return [4 /*yield*/, __await(this.initialize(options))];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        if (!this.hasNext()) return [3 /*break*/, 13];
                        return [4 /*yield*/, __await(this._itemType.read(this._dataStream, {
                                abortSignal: options.abortSignal
                            }))];
                    case 3:
                        result = _b.sent();
                        this._itemsRemainingInBlock--;
                        this._objectIndex++;
                        if (!(this._itemsRemainingInBlock == 0)) return [3 /*break*/, 10];
                        return [4 /*yield*/, __await(AvroParser.readFixedBytes(this._dataStream, AVRO_SYNC_MARKER_SIZE, {
                                abortSignal: options.abortSignal
                            }))];
                    case 4:
                        marker = _b.sent();
                        this._blockOffset = this._initialBlockOffset + this._dataStream.position;
                        this._objectIndex = 0;
                        if (!arraysEqual(this._syncMarker, marker)) {
                            throw new Error("Stream is not a valid Avro file.");
                        }
                        _b.label = 5;
                    case 5:
                        _b.trys.push([5, 7, , 8]);
                        _a = this;
                        return [4 /*yield*/, __await(AvroParser.readLong(this._dataStream, {
                                abortSignal: options.abortSignal
                            }))];
                    case 6:
                        _a._itemsRemainingInBlock = _b.sent();
                        return [3 /*break*/, 8];
                    case 7:
                        err_1 = _b.sent();
                        // We hit the end of the stream.
                        this._itemsRemainingInBlock = 0;
                        return [3 /*break*/, 8];
                    case 8:
                        if (!(this._itemsRemainingInBlock > 0)) return [3 /*break*/, 10];
                        // Ignore block size
                        return [4 /*yield*/, __await(AvroParser.readLong(this._dataStream, { abortSignal: options.abortSignal }))];
                    case 9:
                        // Ignore block size
                        _b.sent();
                        _b.label = 10;
                    case 10: return [4 /*yield*/, __await(result)];
                    case 11: return [4 /*yield*/, _b.sent()];
                    case 12:
                        _b.sent();
                        return [3 /*break*/, 2];
                    case 13: return [2 /*return*/];
                }
            });
        });
    };
    return AvroReader;
}());
export { AvroReader };
