// @azure/core-http/es/src/util/exponentialBackoffStrategy.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export var DEFAULT_CLIENT_RETRY_COUNT = 3;
// intervals are in ms
export var DEFAULT_CLIENT_RETRY_INTERVAL = 1000 * 30;
export var DEFAULT_CLIENT_MAX_RETRY_INTERVAL = 1000 * 90;
export var DEFAULT_CLIENT_MIN_RETRY_INTERVAL = 1000 * 3;
export function isNumber(n) {
    return typeof n === "number";
}
/**
 * @internal
 * Determines if the operation should be retried.
 *
 * @param {number} retryLimit Specifies the max number of retries.
 * @param {(response?: HttpOperationResponse, error?: RetryError) => boolean} predicate Initial chekck on whether to retry based on given responses or errors
 * @param {RetryData} retryData  The retry data.
 * @return {boolean} True if the operation qualifies for a retry; false otherwise.
 */
export function shouldRetry(retryLimit, predicate, retryData, response, error) {
    if (!predicate(response, error)) {
        return false;
    }
    return retryData.retryCount < retryLimit;
}
/**
 * @internal
 * Updates the retry data for the next attempt.
 *
 * @param {RetryPolicyOptions} retryOptions specifies retry interval, and its lower bound and upper bound.
 * @param {RetryData} [retryData]  The retry data.
 * @param {RetryError} [err] The operation"s error, if any.
 */
export function updateRetryData(retryOptions, retryData, err) {
    if (retryData === void 0) { retryData = { retryCount: 0, retryInterval: 0 }; }
    if (err) {
        if (retryData.error) {
            err.innerError = retryData.error;
        }
        retryData.error = err;
    }
    // Adjust retry count
    retryData.retryCount++;
    // Adjust retry interval
    var incrementDelta = Math.pow(2, retryData.retryCount - 1) - 1;
    var boundedRandDelta = retryOptions.retryInterval * 0.8 +
        Math.floor(Math.random() * (retryOptions.retryInterval * 0.4));
    incrementDelta *= boundedRandDelta;
    retryData.retryInterval = Math.min(retryOptions.minRetryInterval + incrementDelta, retryOptions.maxRetryInterval);
    return retryData;
}
