// eslint-plugin-github/lib/configs/flat/recommended.js
const globals = require('globals')
const github = require('../../plugin')
const prettierPlugin = require('eslint-plugin-prettier')
const eslintComments = require('eslint-plugin-eslint-comments')
const importPlugin = require('eslint-plugin-import')
const i18nTextPlugin = require('eslint-plugin-i18n-text')
const noOnlyTestsPlugin = require('eslint-plugin-no-only-tests')
const {fixupPluginRules} = require('@eslint/compat')

module.exports = {
  languageOptions: {
    ecmaVersion: 6,
    sourceType: 'module',
    globals: {
      ...globals.es6,
    },
  },
  plugins: {
    prettierPlugin,
    eslintComments,
    importPlugin,
    'i18n-text': fixupPluginRules(i18nTextPlugin),
    noOnlyTestsPlugin,
    github: fixupPluginRules(github),
  },
  rules: {
    'constructor-super': 'error',
    'eslintComments/disable-enable-pair': 'off',
    'eslintComments/no-aggregating-enable': 'off',
    'eslintComments/no-duplicate-disable': 'error',
    'eslintComments/no-unlimited-disable': 'error',
    'eslintComments/no-unused-disable': 'error',
    'eslintComments/no-unused-enable': 'error',
    'eslintComments/no-use': ['error', {allow: ['eslint', 'eslint-disable-next-line', 'eslint-env', 'globals']}],
    'github/filenames-match-regex': ['error', '^[a-z0-9-]+(.[a-z0-9-]+)?$'],
    'func-style': ['error', 'declaration', {allowArrowFunctions: true}],
    'github/array-foreach': 'error',
    'github/no-implicit-buggy-globals': 'error',
    'github/no-then': 'error',
    'github/no-dynamic-script-tag': 'error',
    'i18n-text/no-en': ['error'],
    'importPlugin/default': 'error',
    'importPlugin/export': 'error',
    'importPlugin/extensions': 'error',
    'importPlugin/first': 'error',
    'importPlugin/named': 'error',
    'importPlugin/namespace': 'error',
    'importPlugin/no-absolute-path': 'error',
    'importPlugin/no-amd': 'error',
    'importPlugin/no-anonymous-default-export': [
      'error',
      {
        allowAnonymousClass: false,
        allowAnonymousFunction: false,
        allowArray: true,
        allowArrowFunction: false,
        allowLiteral: true,
        allowObject: true,
      },
    ],
    'importPlugin/no-commonjs': 'error',
    'importPlugin/no-deprecated': 'error',
    'importPlugin/no-duplicates': 'error',
    'importPlugin/no-dynamic-require': 'error',
    'importPlugin/no-extraneous-dependencies': [0, {devDependencies: false}],
    'importPlugin/no-mutable-exports': 'error',
    'importPlugin/no-named-as-default': 'error',
    'importPlugin/no-named-as-default-member': 'error',
    'importPlugin/no-namespace': 'error',
    'importPlugin/no-unresolved': 'error',
    'importPlugin/no-webpack-loader-syntax': 'error',
    'no-case-declarations': 'error',
    'no-class-assign': 'error',
    'no-compare-neg-zero': 'error',
    'no-cond-assign': 'error',
    'no-console': 'error',
    'no-const-assign': 'error',
    'no-constant-condition': 'error',
    'no-control-regex': 'error',
    'no-debugger': 'error',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty': 'error',
    'no-empty-character-class': 'error',
    'no-empty-pattern': 'error',
    'no-ex-assign': 'error',
    'no-extra-boolean-cast': 'error',
    'no-fallthrough': 'error',
    'no-func-assign': 'error',
    'no-global-assign': 'error',
    'no-implicit-globals': 'error',
    'no-implied-eval': 'error',
    'no-inner-declarations': 'error',
    'no-invalid-regexp': 'error',
    'no-invalid-this': 'error',
    'no-irregular-whitespace': 'error',
    'no-new-symbol': 'error',
    'no-obj-calls': 'error',
    'no-octal': 'error',
    'noOnlyTestsPlugin/no-only-tests': [
      'error',
      {
        block: ['describe', 'it', 'context', 'test', 'tape', 'fixture', 'serial', 'suite'],
      },
    ],
    'no-redeclare': 'error',
    'no-regex-spaces': 'error',
    'no-return-assign': 'error',
    'no-self-assign': 'error',
    'no-sequences': ['error'],
    'no-shadow': 'error',
    'no-sparse-arrays': 'error',
    'no-this-before-super': 'error',
    'no-throw-literal': 'error',
    'no-undef': 'error',
    'no-unreachable': 'error',
    'no-unsafe-finally': 'error',
    'no-unsafe-negation': 'error',
    'no-unused-labels': 'error',
    'no-unused-vars': 'error',
    'no-useless-concat': 'error',
    'no-useless-escape': 'error',
    'no-var': 'error',
    'object-shorthand': ['error', 'always', {avoidQuotes: true}],
    'one-var': ['error', 'never'],
    'prefer-const': 'error',
    'prefer-promise-reject-errors': 'error',
    'prefer-rest-params': 'error',
    'prefer-spread': 'error',
    'prefer-template': 'error',
    'prettierPlugin/prettier': 'error',
    'require-yield': 'error',
    'use-isnan': 'error',
    'valid-typeof': 'error',
    camelcase: ['error', {properties: 'always'}],
    eqeqeq: ['error', 'smart'],
  },
  settings: {
    'importPlugin/resolver': {
      node: {
        extensions: ['.js', '.ts'],
      },
    },
  },
}
