// @azure/storage-blob/dist-esm/storage-blob/src/utils/BlobQuickQueryStream.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { __asyncValues, __awaiter, __extends, __generator } from "tslib";
import { Readable } from "stream";
import { AvroReadableFromStream, AvroReader } from "../../../storage-internal-avro/src";
/**
 * ONLY AVAILABLE IN NODE.JS RUNTIME.
 *
 * A Node.js BlobQuickQueryStream will internally parse avro data stream for blob query.
 *
 * @class BlobQuickQueryStream
 * @extends {Readable}
 */
var BlobQuickQueryStream = /** @class */ (function (_super) {
    __extends(BlobQuickQueryStream, _super);
    /**
     * Creates an instance of BlobQuickQueryStream.
     *
     * @param {NodeJS.ReadableStream} source The current ReadableStream returned from getter
     * @param {BlobQuickQueryStreamOptions} [options={}]
     * @memberof BlobQuickQueryStream
     */
    function BlobQuickQueryStream(source, options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this) || this;
        _this.source = source;
        _this.onProgress = options.onProgress;
        _this.onError = options.onError;
        _this.avroReader = new AvroReader(new AvroReadableFromStream(_this.source));
        _this.avroIter = _this.avroReader.parseObjects({ abortSignal: options.abortSignal });
        return _this;
    }
    BlobQuickQueryStream.prototype._read = function () {
        var _this = this;
        this.readInternal().catch(function (err) {
            _this.emit("error", err);
        });
    };
    BlobQuickQueryStream.prototype.readInternal = function () {
        var e_1, _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b, _c, obj, schema, exit, data, bytesScanned, totalBytes, fatal, name_1, description, position, e_1_1;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 5, 6, 11]);
                        _b = __asyncValues(this.avroIter);
                        _d.label = 1;
                    case 1: return [4 /*yield*/, _b.next()];
                    case 2:
                        if (!(_c = _d.sent(), !_c.done)) return [3 /*break*/, 4];
                        obj = _c.value;
                        schema = obj.$schema;
                        if (typeof schema !== "string") {
                            throw Error("Missing schema in avro record.");
                        }
                        exit = false;
                        switch (schema) {
                            case "com.microsoft.azure.storage.queryBlobContents.resultData":
                                data = obj.data;
                                if (data instanceof Uint8Array === false) {
                                    throw Error("Invalid data in avro result record.");
                                }
                                if (!this.push(Buffer.from(data))) {
                                    exit = true;
                                }
                                break;
                            case "com.microsoft.azure.storage.queryBlobContents.progress":
                                bytesScanned = obj.bytesScanned;
                                if (typeof bytesScanned !== "number") {
                                    throw Error("Invalid bytesScanned in avro progress record.");
                                }
                                if (this.onProgress) {
                                    this.onProgress({ loadedBytes: bytesScanned });
                                }
                                break;
                            case "com.microsoft.azure.storage.queryBlobContents.end":
                                if (this.onProgress) {
                                    totalBytes = obj.totalBytes;
                                    if (typeof totalBytes !== "number") {
                                        throw Error("Invalid totalBytes in avro end record.");
                                    }
                                    this.onProgress({ loadedBytes: totalBytes });
                                }
                                this.push(null);
                                break;
                            case "com.microsoft.azure.storage.queryBlobContents.error":
                                if (this.onError) {
                                    fatal = obj.fatal;
                                    if (typeof fatal !== "boolean") {
                                        throw Error("Invalid fatal in avro error record.");
                                    }
                                    name_1 = obj.name;
                                    if (typeof name_1 !== "string") {
                                        throw Error("Invalid name in avro error record.");
                                    }
                                    description = obj.description;
                                    if (typeof description !== "string") {
                                        throw Error("Invalid description in avro error record.");
                                    }
                                    position = obj.position;
                                    if (typeof position !== "number") {
                                        throw Error("Invalid position in avro error record.");
                                    }
                                    this.onError({
                                        position: position,
                                        name: name_1,
                                        isFatal: fatal,
                                        description: description
                                    });
                                }
                                break;
                            default:
                                throw Error("Unknown schema " + schema + " in avro progress record.");
                        }
                        if (exit) {
                            return [3 /*break*/, 4];
                        }
                        _d.label = 3;
                    case 3: return [3 /*break*/, 1];
                    case 4: return [3 /*break*/, 11];
                    case 5:
                        e_1_1 = _d.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 11];
                    case 6:
                        _d.trys.push([6, , 9, 10]);
                        if (!(_c && !_c.done && (_a = _b.return))) return [3 /*break*/, 8];
                        return [4 /*yield*/, _a.call(_b)];
                    case 7:
                        _d.sent();
                        _d.label = 8;
                    case 8: return [3 /*break*/, 10];
                    case 9:
                        if (e_1) throw e_1.error;
                        return [7 /*endfinally*/];
                    case 10: return [7 /*endfinally*/];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    return BlobQuickQueryStream;
}(Readable));
export { BlobQuickQueryStream };
